### Part A — Understand the Data

In [1]:
import pandas as pd

In [2]:
df=pd.read_csv("Sales_Dataset.csv")

In [3]:
df.head(10)

,Order_ID,Date,Product,Category,Quantity,Price,City
0,1001,2026-01-11,Mobile,Electronics,3.0,42750.0,Lahore
1,1002,2026-01-31,Headphones,Electronics,5.0,6175.0,Peshawar
2,1003,2026-01-13,Printer,Office,1.0,26600.0,Lahore
3,1004,2026-05-14,Keyboard,Accessories,5.0,4275.0,Peshawar
4,1005,2026-06-20,Keyboard,Accessories,5.0,4500.0,Lahore
5,1006,2026-06-04,Notebook,Stationery,3.0,475.0,Lahore
6,1007,2026-04-02,Printer,Office,3.0,26600.0,Lahore
7,1008,2026-01-31,Monitor,Electronics,NaN,32000.0,Islamabad
8,1009,2026-04-03,Monitor,Electronics,5.0,32000.0,Islamabad
9,1010,2026-05-22,Notebook,Stationery,1.0,500.0,Islamabad


In [4]:
df.tail(5)

,Order_ID,Date,Product,Category,Quantity,Price,City
57,1058,2026-03-01,Mouse,Accessories,1.0,2625.0,Multan
58,1059,2026-03-26,Laptop,Electronics,1.0,80750.0,Peshawar
59,1060,2026-05-13,Notebook,Stationery,5.0,475.0,Islamabad
60,1015,2026-02-27,Monitor,Electronics,3.0,33600.0,Multan
61,1038,2026-02-28,Printer,Office,5.0,29400.0,Multan


In [5]:
df.shape

(62, 7)

In [6]:
df.columns

Index(['Order_ID', 'Date', 'Product', 'Category', 'Quantity', 'Price', 'City'], dtype='object')

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 62 entries, 0 to 61
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Order_ID  62 non-null     int64  
 1   Date      62 non-null     object 
 2   Product   62 non-null     object 
 3   Category  62 non-null     object 
 4   Quantity  61 non-null     float64
 5   Price     61 non-null     float64
 6   City      62 non-null     object 
dtypes: float64(2), int64(1), object(4)
memory usage: 3.5+ KB


In [8]:
df.describe()

,Order_ID,Quantity,Price
count,62.000000,61.000000,61.000000
mean,1030.370968,3.163934,23219.262295
std,17.315996,1.428056,25118.734457
min,1001.000000,1.000000,475.000000
25%,1015.250000,2.000000,2625.000000
50%,1030.500000,3.000000,6175.000000
75%,1044.750000,4.000000,42750.000000
max,1060.000000,5.000000,85000.000000


In [9]:
df.isnull().sum()

Order_ID    0
Date        0
Product     0
Category    0
Quantity    1
Price       1
City        0
dtype: int64

In [10]:
df.duplicated().sum()

np.int64(2)

### Part B — Data Cleaning

In [11]:
df["Date"]=pd.to_datetime(df["Date"])

In [12]:
df["OrderYear"]=df["Date"].dt.year

In [13]:
df["OrderMonth"]=df["Date"].dt.month

In [14]:
df["OrderDay"]=df["Date"].dt.day

In [15]:
df

,Order_ID,Date,Product,Category,Quantity,Price,City,OrderYear,OrderMonth,OrderDay
0,1001,2026-01-11,Mobile,Electronics,3.0,42750.0,Lahore,2026,1,11
1,1002,2026-01-31,Headphones,Electronics,5.0,6175.0,Peshawar,2026,1,31
2,1003,2026-01-13,Printer,Office,1.0,26600.0,Lahore,2026,1,13
3,1004,2026-05-14,Keyboard,Accessories,5.0,4275.0,Peshawar,2026,5,14
4,1005,2026-06-20,Keyboard,Accessories,5.0,4500.0,Lahore,2026,6,20
...,...,...,...,...,...,...,...,...,...,...
57,1058,2026-03-01,Mouse,Accessories,1.0,2625.0,Multan,2026,3,1
58,1059,2026-03-26,Laptop,Electronics,1.0,80750.0,Peshawar,2026,3,26
59,1060,2026-05-13,Notebook,Stationery,5.0,475.0,Islamabad,2026,5,13
60,1015,2026-02-27,Monitor,Electronics,3.0,33600.0,Multan,2026,2,27


In [16]:
df[df.isnull().any(axis=1)]

,Order_ID,Date,Product,Category,Quantity,Price,City,OrderYear,OrderMonth,OrderDay
7,1008,2026-01-31,Monitor,Electronics,NaN,32000.0,Islamabad,2026,1,31
22,1023,2026-06-03,Printer,Office,4.0,NaN,Lahore,2026,6,3


In [17]:
df[df["Product"]=="Monitor"]

,Order_ID,Date,Product,Category,Quantity,Price,City,OrderYear,OrderMonth,OrderDay
7,1008,2026-01-31,Monitor,Electronics,NaN,32000.0,Islamabad,2026,1,31
8,1009,2026-04-03,Monitor,Electronics,5.0,32000.0,Islamabad,2026,4,3
14,1015,2026-02-27,Monitor,Electronics,3.0,33600.0,Multan,2026,2,27
28,1029,2026-02-02,Monitor,Electronics,3.0,32000.0,Lahore,2026,2,2
39,1040,2026-04-27,Monitor,Electronics,5.0,32000.0,Islamabad,2026,4,27
60,1015,2026-02-27,Monitor,Electronics,3.0,33600.0,Multan,2026,2,27


In [18]:
df[df["Product"]=="Printer"]

,Order_ID,Date,Product,Category,Quantity,Price,City,OrderYear,OrderMonth,OrderDay
2,1003,2026-01-13,Printer,Office,1.0,26600.0,Lahore,2026,1,13
6,1007,2026-04-02,Printer,Office,3.0,26600.0,Lahore,2026,4,2
22,1023,2026-06-03,Printer,Office,4.0,NaN,Lahore,2026,6,3
37,1038,2026-02-28,Printer,Office,5.0,29400.0,Multan,2026,2,28
47,1048,2026-05-04,Printer,Office,1.0,29400.0,Multan,2026,5,4
56,1057,2026-05-27,Printer,Office,4.0,26600.0,Lahore,2026,5,27
61,1038,2026-02-28,Printer,Office,5.0,29400.0,Multan,2026,2,28


In [19]:
df.groupby("Product")["Price"].mean()

Product
Headphones     6391.666667
Keyboard       4413.461538
Laptop        83300.000000
Mobile        45187.500000
Monitor       32533.333333
Mouse          2500.000000
Notebook        488.636364
Printer       28000.000000
Name: Price, dtype: float64

In [20]:
df.groupby("Product")["Quantity"].mean()

Product
Headphones    4.666667
Keyboard      2.846154
Laptop        2.400000
Mobile        3.250000
Monitor       3.800000
Mouse         2.400000
Notebook      3.363636
Printer       3.285714
Name: Quantity, dtype: float64

In [21]:
df["Price"]=df["Price"].fillna(26600.0)

The price charged for the printer in different month for Lahore city is samei.e 26600.0.

(Same price(29400.0)is charged for Multan across different orders) .So i put the same value in this for Lahore.

In [22]:
df["Quantity"]=df["Quantity"].fillna(3.0) #fill with the median value

"The missing Quantity was filled with the median quantity because quantity is numerical and the median is less sensitive to extreme values."

In [23]:
df[df.duplicated()]

,Order_ID,Date,Product,Category,Quantity,Price,City,OrderYear,OrderMonth,OrderDay
60,1015,2026-02-27,Monitor,Electronics,3.0,33600.0,Multan,2026,2,27
61,1038,2026-02-28,Printer,Office,5.0,29400.0,Multan,2026,2,28


In [24]:
df=df.drop_duplicates()

In [25]:
df.dtypes

Order_ID               int64
Date          datetime64[ns]
Product               object
Category              object
Quantity             float64
Price                float64
City                  object
OrderYear              int32
OrderMonth             int32
OrderDay               int32
dtype: object

In [26]:
df.isnull().sum()

Order_ID      0
Date          0
Product       0
Category      0
Quantity      0
Price         0
City          0
OrderYear     0
OrderMonth    0
OrderDay      0
dtype: int64

In [27]:
df.to_csv("Sales_Data_Cleaned", index=False)

### Part C — Feature Engineering

In [28]:
df=df.copy()

In [29]:
df["Revenue"]=df["Quantity"]*df["Price"]

### Part D — Exploratory Questions

In [30]:
#Q12 Total Revenue
df["Revenue"].sum()

np.float64(4180225.0)

### Total Revenue is the sum of sales of all the products i.e 4180225.0)

In [31]:
#Q13 Average order revenue
df["Revenue"].mean()

np.float64(69670.41666666667)

### Average order revenue is the revenue/per order i.e 69670.41

In [32]:
#Q14 Revenue by Product
df.groupby("Product")["Revenue"].sum()

Product
Headphones      89700.0
Keyboard       163575.0
Laptop        1011500.0
Mobile        1759500.0
Monitor        612800.0
Mouse           29625.0
Notebook        17925.0
Printer        495600.0
Name: Revenue, dtype: float64

### Mobiles generate the highest revenue i.e1 759500.0

In [33]:
df.groupby("City")["Revenue"].sum()

City
Islamabad     1393625.0
Karachi        587800.0
Lahore         590875.0
Multan         730325.0
Peshawar       387075.0
Rawalpindi     490525.0
Name: Revenue, dtype: float64

### Highest revenue generated from Islamabad.

In [34]:
df.groupby("Category")["Revenue"].sum()

Category
Accessories     193200.0
Electronics    3473500.0
Office          495600.0
Stationery       17925.0
Name: Revenue, dtype: float64

### Electronics generated the highest revenue.

In [35]:
df.groupby("Product")["Quantity"].sum()

Product
Headphones    14.0
Keyboard      37.0
Laptop        12.0
Mobile        39.0
Monitor       19.0
Mouse         12.0
Notebook      37.0
Printer       18.0
Name: Quantity, dtype: float64

### Mobile phones were sold in Highest quantity

In [36]:
df.groupby("City")["Revenue"].mean()

City
Islamabad     116135.416667
Karachi        73475.000000
Lahore         45451.923077
Multan         60860.416667
Peshawar       43008.333333
Rawalpindi     81754.166667
Name: Revenue, dtype: float64

### Islamabad has Highest average order revenue.

In [37]:
df.groupby("OrderMonth")["Revenue"].sum()

OrderMonth
1    877900.0
2    491400.0
3    649650.0
4    978225.0
5    302600.0
6    880450.0
Name: Revenue, dtype: float64

### April generated the highest monthly revenue of Rs. 978,225.

In [38]:
df.sort_values(by="Revenue",ascending=False).head(5)

,Order_ID,Date,Product,Category,Quantity,Price,City,OrderYear,OrderMonth,OrderDay,Revenue
18,1019,2026-03-26,Laptop,Electronics,4.0,85000.0,Islamabad,2026,3,26,340000.0
32,1033,2026-06-07,Laptop,Electronics,3.0,85000.0,Islamabad,2026,6,7,255000.0
55,1056,2026-04-14,Laptop,Electronics,3.0,85000.0,Karachi,2026,4,14,255000.0
27,1028,2026-06-28,Mobile,Electronics,5.0,45000.0,Multan,2026,6,28,225000.0
52,1053,2026-01-17,Mobile,Electronics,5.0,42750.0,Islamabad,2026,1,17,213750.0


### Top 5 orders include the 10 laptop and 10 mobiles generating the max revenue

### Part E — Business Questions 💼

Q21

Which product should the company focus on based on revenue?

Explain why.


Company should focus on the market for mobile phones and laptops as these products contribute more in the total revenue. 

Follows is the evidence that max revenue has been generating from the sales of mobiles and laptops. 

Laptop 1011500.0 
Mobile 1759500.0



Q22

Which city appears to be the strongest market?

Use your analysis to support your answer.


As Islamabad generates the highest total revenue it seems to be the strongest market.

Islamabad     1393625.0
Karachi        587800.0
Lahore         590875.0
Multan         730325.0
Peshawar       387075.0
Rawalpindi     490525.0

Q23

Which category needs attention?

Don't just say which one—give evidence.

Lowest revenue generating categories (including Office,Stationary and Accessories) should be
be given more attention to increase the sales and revenue.

Accessories     193200.0

Electronics    3473500.0

Office          495600.0

Stationery       17925.0


"Accessories and Stationery generate relatively low revenue and may need further investigation."

#Q24 

Mobile has the highest quantity sold (39 units) and also the highest revenue (Rs. 1.76 million).

Islamabad has both the highest total revenue and the highest average order revenue.

### EDA Challenge

Sales Analysis Summary

Total Revenue: 4180225.0

Average Order Revenue: 69670.41

Best-Selling Product:Mobile Phones

Highest-Revenue Product: Mobile Phones

Best City: Islamabad

Best Category: Electronics

Key Insights:

1. Electronics contribute more in the revenue generation.
2. Highest selling product is mobile phones generating the max revenue.
3. The best market that generate the max revenue is Islamabad.
4. In the April, revenues is highest than other months.

Business Recommendation:
Lowest revenue generating categories (including Office,Stationary and Accessories) should be be given more attention to increase the sales and revenue.Other markets should be focused to increase sales in other cities through promotional strategies.

